addons/binding/org.openhab.binding.upb/src/main/java/org/openhab/binding/upb/internal/UPBWriter.java

```diff
@@ -0,0 +1,205 @@
+
+    /**
+     * The UPB modem's OutputStream.
+     */
+    private OutputStream outputStream;
+
+    /**
+     * UPBReader that is monitoring the modem's InputStream.
+     */
+    private UPBReader upbReader;
+
+    /**
+     * Instantiates a new {@link UPBWriter} using the given modem
+     * {@link OutputStream}.
+     *
+     * @param outputStream
+     *            the {@link OutputStream} from the UPB modem.
+     * @param upbReader
+     *            the {@link UPBReader} that is monitoring the same UPB modem.
+     */
+    public UPBWriter(OutputStream outputStream, UPBReader upbReader) {
+        this.outputStream = outputStream;
+        this.upbReader = upbReader;
+    }
+
+    /**
+     * Queues a message to be written to the modem.
+     *
+     * @param message
+     *            the message to be written.
+     */
+    public void queueMessage(MessageBuilder message) {
+        String data = message.build();
+        logger.debug("Queueing message {}.", data);
+        executor.execute(new FutureMessage(new Message(data.getBytes(), message.getPriority())));
+    }
+
+    /**
+     * Cancels all queued messages and releases resources. This instance cannot
+     * be used again and a new {@link UPBWriter} must be instantiated after
+     * calling this method.
+     */
+    public void shutdown() {
+        executor.shutdownNow();
+
+        try {
+            outputStream.close();
+        } catch (IOException e) {
+        }
+        logger.debug("UPBWriter shutdown");
+    }
+
+    private static class FutureMessage extends FutureTask<FutureMessage> implements Comparable<FutureMessage> {
+        private Message message;
+
+        public FutureMessage(Message message) {
+            super(message, null);
+
+            this.message = message;
+        }
+
+        @Override
+        public int compareTo(FutureMessage o) {
+            return message.priority.getValue() - o.message.priority.getValue();
+        }
+    }
+
+    /**
+     * {@link Runnable} implementation used to write data to the UPB modem.
+     *
+     * @author cvanorman
+     *
+     */
+    private class Message implements Runnable, Listener {
+
+        private boolean waitingOnAck = true;
+        private boolean ackReceived = false;
+
+        private byte[] data;
+        private UPBMessage.Priority priority;
+
+        private Message(byte[] data, UPBMessage.Priority priority) {
+            this.data = data;
+
+            this.priority = priority;
+        }
+
+        private synchronized void ackReceived(boolean ack) {
+            waitingOnAck = false;
+            ackReceived = ack;
+            notify();
+        }
+
+        private synchronized boolean waitForAck(int retryCount) {
+            long start = System.currentTimeMillis();
+            while (waitingOnAck && (System.currentTimeMillis() - start) < ACK_TIMEOUT) {
+                try {
+                    wait(ACK_TIMEOUT);
+                } catch (InterruptedException e) {
+
```

Comment:
If we receive this InterruptedException, the interrupted flag was cleared and we most likely need to reset the flag:

http://www.yegor256.com/2015/10/20/interrupted-exception.html
https://www.ibm.com/developerworks/library/j-jtp05236/